neo4j-to-neptune/src/main/java/com/amazonaws/services/neptune/util/NeptuneBulkLoader.java

```diff
@@ -294,34 +279,43 @@ protected CompletableFuture<Boolean> uploadFileWithInflightCompression(String lo
             System.err.println("Initiating Transfer Manager upload...");
             Upload upload = transferManager.upload(uploadRequest);
 
-            // Wait for BOTH upload and compression to complete - fail if either fails
             return CompletableFuture.allOf(upload.completionFuture(), compressionFuture)
-                .thenApply(ignored -> {
-                    System.err.println(
-                        "Successfully uploaded " + localFile.getName() +
-                        " (compressed) - ETag: " + upload.completionFuture().join().response().eTag());
-                    return true;
-                })
-                .exceptionally(throwable -> {
-                    logUploadError(localFilePath, throwable);
-                    // Re-throw to maintain fail-fast behavior
-                    if (throwable instanceof RuntimeException) {
-                        throw (RuntimeException) throwable;
-                    } else {
-                        throw new RuntimeException("Upload or compression failed", throwable);
-                    }
-                })
-                .whenComplete((result, throwable) -> {
-                    closeStreams(streamExecutor, pipedOut, pipedIn);
-                });
-
+                .thenApply(ignored -> handleUploadSuccess(localFile, upload))
+                .exceptionally(throwable -> handleUploadFailure(localFilePath, throwable))
+                .whenComplete((result, throwable) -> closeStreams(streamExecutor, pipedOut, pipedIn));
         } catch (Exception e) {
```

Comment:
You should be able to handle the cleanup with a `finally` block.
```suggestion
        } finally {
```

Reply:
Refactored how the method does the error catching. It now just throws it up to the caller method and has all of the logging and exception checking done in there.

Removed the catch block and added a finally to close off the thread & streams